### PR TITLE
Bugfix: Ensure multiple runs invariant

### PR DESCRIPTION
**Pull request by mr-archano**
## Scope of the PR
As @ouchadam mentioned in issue #11:
> Given a previously failed result, when running the task again the build does not fail.

This shouldn't happen because code is identical, therefore should yield the same results. As already mentioned in #11, after a quick investigation:
> The problem seems related to the fact the violations are now collected only if static analysis tasks (eg: `checkstyleMain`) are run. When one of these tasks is `UP-TO-DATE` because inputs are the same then we don't evaluate the reports (already generated) and this yields a wrong result for `evaluateViolations`.

## Considerations and implementation
Each analysis task (`Checkstyle`, `Findbugs`, `Pmd`) is now depending on a specific implementation of `CollectionViolationsTask` that will perform the same action before provided as `doLast`. This will allow to evaluate the reports and collect the violations even when the analysis are skipped because their input haven't changed.

### Test(s) added 
For each tool supported by the plugin an integration test has been added to ensure that 2 consecutive runs without any change in the code yield the same result.

### CODE REVIEW

File: plugin/src/main/groovy/com/novoda/staticanalysis/internal/CollectViolationsTask.groovy [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    CollectViolationsTask() {
+        onlyIf { xmlReportFile?.exists() }
+    }
+
+    void setXmlReportFile(File xmlReportFile) {
+        this.xmlReportFile = xmlReportFile
+    }
+
+    void setViolations(Violations violations) {
```
Comment (blundell) [inline plugin/src/main/groovy/com/novoda/staticanalysis/internal/CollectViolationsTask.groovy]: heh this reminds me of the PMD source code

File: plugin/src/main/groovy/com/novoda/staticanalysis/internal/CollectViolationsTask.groovy [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    final void run() {
+        collectViolations(xmlReportFile, htmlReportFile, violations)
+    }
+
+    File getXmlReportFile() {
+        return xmlReportFile
+    }
+
+    File getHtmlReportFile() {
+        new File(xmlReportFile.absolutePath - '.xml' + '.html')
```
Comment (blundell) [inline plugin/src/main/groovy/com/novoda/staticanalysis/internal/CollectViolationsTask.groovy]: cool syntax

File: plugin/src/test/groovy/com/novoda/staticanalysis/internal/checkstyle/CheckstyleIntegrationTest.groovy [hunk 3]
```diff
@@ -44,6 +44,29 @@ public class CheckstyleIntegrationTest {
     }
 
     @Test
+    public void shouldFailBuildAgainWhenCheckstyleWarningsStillOverTheThresholdAfterSecondRun() {
```
Comment (blundell) [inline plugin/src/test/groovy/com/novoda/staticanalysis/internal/checkstyle/CheckstyleIntegrationTest.groovy]: 👍 

might be nice to word it differently tho

`afterSecondRunWhenStillOverTheThresholdThenTheBuildFailsAgain`

or if you insist on starting with should:

`shouldFailTheBuildAfterSecondRunWhenStillOverTheThreshold`

File: plugin/src/main/groovy/com/novoda/staticanalysis/internal/pmd/CollectPmdViolationsTask.groovy [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+import com.novoda.staticanalysis.internal.CollectViolationsTask
+import com.novoda.staticanalysis.internal.Violations
+
+class CollectPmdViolationsTask extends CollectViolationsTask {
+
+    @Override
+    void collectViolations(File xmlReportFile, File htmlReportFile, Violations violations) {
+        PmdViolationsEvaluator evaluator = new PmdViolationsEvaluator(xmlReportFile)
+        int errors = 0, warnings = 0
```
Comment (mr-archano) [inline plugin/src/main/groovy/com/novoda/staticanalysis/internal/pmd/CollectPmdViolationsTask.groovy]: LoL, I tried.

File: plugin/src/main/groovy/com/novoda/staticanalysis/internal/findbugs/GenerateFindBugsHtmlReport.groovy [hunk 5]
```diff
@@ -0,0 +1,19 @@
+package com.novoda.staticanalysis.internal.findbugs
+
+import org.gradle.api.tasks.JavaExec
+
+class GenerateFindBugsHtmlReport extends JavaExec {
```
Comment (ouchadam) [inline plugin/src/main/groovy/com/novoda/staticanalysis/internal/findbugs/GenerateFindBugsHtmlReport.groovy]: 👍

**Comment by ouchadam** [comment]
nice set of tests for each plugin, LGTM ! 👍

**Pull request merged**